File: 1289-minimum-falling-path-sum-ii/1289-minimum-falling-path-sum-ii.py
class Solution(object):
    def minFallingPathSum(self, grid):
        n = len(grid)
        dp = grid[n - 1]

        for i in range(n - 2, -1, -1):
            first_min, second_min, idx = float('inf'), float('inf'), -1
            for j in range(n):
                if first_min > dp[j]:
                    second_min = first_min
                    first_min = dp[j]
                    idx = j
                elif second_min > dp[j]:
                    second_min = dp[j]
            
            for j in range(n):
                if j == idx:
                    dp[j] = grid[i][j] + second_min
                else:
                    dp[j] = grid[i][j] + first_min
            
        return min(dp)
        

File: 1289-minimum-falling-path-sum-ii/1289-minimum-falling-path-sum-ii.java
class Solution {
    public int minFallingPathSum(int[][] grid) {
        int n = grid.length;
        int[] dp = new int[n];

        for(int i = 0; i<n; i++) {
            dp[i] = grid[n-1][i];
        }

        for(int i = n - 2; i>=0; i--) {
            int firstMin = Integer.MAX_VALUE, secondMin = Integer.MAX_VALUE, idx = -1;
            for(int j = 0; j<n; j++) {
                if(firstMin > dp[j]) {
                    secondMin = firstMin;
                    firstMin = dp[j];
                    idx = j;
                }
                else if(secondMin > dp[j]) {
                    secondMin = dp[j];
                }
            }

            for(int j = 0; j<n; j++) {
                if(j == idx) dp[j] = secondMin + grid[i][j];
                else dp[j] = firstMin + grid[i][j];
            }
        }

        int ans = Integer.MAX_VALUE;
        for(int i = 0; i<n; i++) {
            ans = Math.min(ans, dp[i]);
        }

        return ans;
    }
}